packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -1066,14 +1070,35 @@ void Add(
     prefix << "Failed to compute Op(B). Please report this bug to the Tpetra developers.");
 #endif // HAVE_TPETRA_DEBUG
 
+  bool CwasFillComplete = false;
+
   // Allocate or zero the entries of the result matrix.
   if (! C.is_null ()) {
+    CwasFillComplete = C->isFillComplete();
+    if(CwasFillComplete)
+      C->resumeFill();
     C->setAllToScalar (STS::zero ());
   } else {
     // FIXME (mfh 08 May 2013) When I first looked at this method, I
     // noticed that C was being given the row Map of Aprime (the
     // possibly transposed version of A).  Is this what we want?
-    C = rcp (new crs_matrix_type (Aprime->getRowMap (), 0));
+
+    // It is a precondition that Aprime and Bprime have the same domain and range maps.
+    // However, they may have different row maps. In this case, it's difficult to
+    // get a precise upper bound on the number of entries in each local row of C, so
+    // just use the looser upper bound based on the max number of entries in any row of Aprime and Bprime.
+    if(Aprime->getRowMap()->isSameAs(*Bprime->getRowMap())) {
+      LocalOrdinal numLocalRows = Aprime->getLocalNumRows();
+      Array<size_t> CmaxEntriesPerRow(numLocalRows);
+      for(LocalOrdinal i = 0; i < numLocalRows; i++) {
+        CmaxEntriesPerRow[i] = Aprime->getNumEntriesInLocalRow(i) + Bprime->getNumEntriesInLocalRow(i);
+      }
+      C = rcp (new crs_matrix_type (Aprime->getRowMap (), CmaxEntriesPerRow()));
+    }
+    else {
+      // Make sure Aprime and Bprime have global constants, before we ask for max entries per row.
```

Comment:
How is this enforced?

Reply:
It's asserted up on lines 1013 and 1016.